include/maths/analytics/CBoostedTreeUtils.h

```diff
@@ -50,37 +96,132 @@ enum EExtraColumnTag {
+//! so we can always safely load four values at once and to maintain the spacing.
+//! In total though this only adds up to 16 bytes overhead.
+//!
+//! How does this compare to std::upper_bound performance-wise?
+//!
+//! The following are representative figures for bare metal for 10000000 lookups:
+//!
+//!  collection size | std::upper_bound | CSearchTree::upperBound | speedup
+//!  --------------- | ---------------- | ----------------------- | -------
+//!        100       |      215 ms      |          44 ms          |  4.8 X
+//!       10000      |      580 ms      |          84 ms          |  6.9 X
+//!
+//! One might reasonably expect larger speedups for larger data set sizes because
+//! of the higher branch factor. We posit that one pays fixed overheads due to
+//! cache misses traversing larger data sets and this reduces the % improve from
+//! executing fewer instructions.
+class MATHS_ANALYTICS_EXPORT CSearchTree {
+public:
+    CSearchTree() = default;
+    explicit CSearchTree(const TFloatVec& values);
+
+    //! Check if it's empty.
+    bool empty() const { return m_Size == 0; }
+
+    //! Get the number of items in the set.
+    std::size_t size() const { return m_Size; }
+
+    //! A drop in replacement for std::upper_bound on a sorted collection.
+    std::size_t upperBound(common::CFloatStorage x) const {
+        // These branches should be predictably false most of the time for our
+        // usage and so almost free.
+        if (m_Size == 0 || x < m_Min) {
+            return 0;
+        }
+        if (x >= INF) {
+            return m_Size;
+        }
+
+        std::size_t node{0};
+        std::size_t offset{0};
+        auto vecx = ml_broadcast_load_128(&x.cstorage());
+
+        for (auto branchSize : m_BranchSizes) {
+            std::size_t branch{selectBranch(&m_Values[node], vecx)};
+            LOG_TRACE(<< "node = " << node << "/" << this->printNode(node)
+                      << ", branch = " << branch);
+
+            // Note that node is a multiple of 4. This follows from the fact that
+            // the step size 4 + 5^n - 1 is a multiple of 4 which can be shown by
+            // induction:
+            //
+            //   5^n - 1 = 5 * (5^(n - 1) - 1) + 4 and 5 - 1 = 1 * 4.
+            //
+            // This means that since m_Values are 16 byte aligned the values at node
+            // are 16 byte aligned and we can safely read them using an aligned load.
+            node += 4 + (branchSize - 1) * branch;
+
+            // Each branch point which is greater than x is out of order w.r.t. this
+            // point and must be subtracted from node to get the correct upper bound.
+            offset += 4 - branch;
+        }
+
+        std::size_t branch{selectBranch(&m_Values[node], vecx)};
+        LOG_TRACE(<< "x = " << x << ", node = " << node << "/" << this->printNode(node)
+                  << ", branch = " << branch << ", offset = " << offset);
+
+        return node + branch + 1 - offset;
```

Comment:
The variables `node` and `branch` refer to the tree representation, while `offset` refers to the original array. It may create a confusion. It would be great if you could make the names more explicit.

Reply:
See [8f1c5d3](https://github.com/elastic/ml-cpp/pull/2380/commits/8f1c5d3618a84255bf44907364e0838d05e3ee5b).